oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -0,0 +1,349 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonParser;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * AWS credentials representing a third-party identity for calling Google APIs.
+ *
+ * <p>By default, attempts to exchange the external credential for a GCP access token.
+ */
+public class AwsCredentials extends ExternalAccountCredentials {
+
+  /**
+   * The AWS credential source. Stores data required to retrieve the AWS credential from the AWS
+   * metadata server.
+   */
+  static class AwsCredentialSource extends CredentialSource {
+
+    private final String regionUrl;
+    private final String url;
+    private final String regionalCredentialVerificationUrl;
+
+    /**
+     * The source of the AWS credential. The credential source map must contain the
+     * `regional_cred_verification_url` field.
+     *
+     * <p>The `regional_cred_verification_url` is the regional GetCallerIdentity action URL, used to
+     * determine the account ID and its roles.
+     *
+     * <p>The `environment_id` is the environment identifier, in the format “aws${version}”. This
+     * indicates whether breaking changes were introduced to the underlying AWS implementation.
+     *
+     * <p>The `region_url` identifies the targeted region. Optional.
+     *
+     * <p>The `url` locates the metadata server used to retrieve the AWS credentials. Optional.
+     */
+    AwsCredentialSource(Map<String, Object> credentialSourceMap) {
+      super(credentialSourceMap);
+      if (!credentialSourceMap.containsKey("regional_cred_verification_url")) {
+        throw new IllegalArgumentException(
+            "A regional_cred_verification_url representing the"
+                + " GetCallerIdentity action URL must be specified.");
+      }
+
+      String environmentId = (String) credentialSourceMap.get("environment_id");
+
+      // Environment version is prefixed by "aws". e.g. "aws1".
+      Matcher matcher = Pattern.compile("^(aws)([\\d]+)$").matcher(environmentId);
+      if (!matcher.find()) {
```

Comment:
Is this a multiline string? If not, you probably just want matches() here instead of find(), and you don't need to bind to the beginning and end of the line.